Document errors (#314)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

- session_key, the shared session secret.

Exceptions:

> - InvalidServerAkeMACError, the handshake fails.

Reviewer: Can we rename this to "HandshakeAuthenticationError", "ServerAuthenticationFailure," or similar? I don't know if we need to note "AKE" and "MAC".

Author: Isn't that more telling?
I fear that ServerAuthenticationFailure is too vague and doesn't say what the actual error is. e.g. it could be mistaken with the server sending the wrong public key (or whatever - this would result in the same error). I like the precision of this wording to pinpoint what went wrong.

Reviewer: Maybe ServerAuthenticationMACError then?

a production deployment of OPAQUE: for instance, ensuring that there is a
server-authenticated channel over which OPAQUE registration and login is run.

# API Considerations

## Errors {#api-errors}

This specification based on OPRF and 3DH only has a couple of fallible functions. For example,

> MACs exchanged during the AKE that might don't verify. Additionally, given the statically sized nature

Reviewer 2: ```suggestion
MACs exchanged during the AKE that might not verify. Additionally, given the statically sized nature
```
Not exactly sure, english isn't my mother tongue.

Author: Mine neither :) but you're correct !

a production deployment of OPAQUE: for instance, ensuring that there is a
server-authenticated channel over which OPAQUE registration and login is run.

# API Considerations

## Errors {#api-errors}

This specification based on OPRF and 3DH only has a couple of fallible functions. For example,

> MACs exchanged during the AKE that might not verify. Additionally, given the statically sized nature
> of the protocol messages, clients and servers can easily detect errors in invalid messages or group elements.

Reviewer: ```suggestion
MACs exchanged during the AKE might fail verification, yielding an error. Additionally, protocol messages 
sent between client and server might not match their expected size, yielding a different error.
```

| ErrInvalidClientEPK     | Server | Ke1                       | Invalid ephemeral client public key.                 |
| ErrInvalidServerEPK     | Client | Ke2                       | Invalid ephemeral server public key.                 |

### Client errors {#client-errors}

The client can produce errors due to incorrect values in the messages it received, either because of
invalid data, point encoding, or MACs.

| Name                   | Stage        | Reason                                                                   |
|:-----------------------|:-------------|:-------------------------------------------------------------------------|
| ErrInvalidMaskedLength | ClientFinish | The length of the masked response is not = point length + envelope size. |
| ErrInvalidPKS          | ClientFinish | Invalid server public key in unmasked response.                          |
| ErrEnvelopeInvalidMac  | ClientFinish | Invalid envelope authentication tag.                                     |
| ErrAkeInvalidServerMac | ClientFinish | Invalid server AKE MAC.                                                  |

### Server errors {#server-errors}

> The server returns a protocol error in the sole case of an invalid client MAC in the AKE. 

Reviewer: ```suggestion
The following table enumerates server errors that can occur, where they occur, and the reason for the error.
In this case, servers can only generate a single protocol error.
```

| Name                   | Stage         | Reason                  |
|:-----------------------|:--------------|:------------------------|
| ErrAkeInvalidClientMac | ServerFinish  | Invalid client AKE MAC. |

### Application errors

Implementations MUST verify certain input arguments and return an appropriate error. Note that

> these are not protocol errors, and are indicative.

Reviewer: ```suggestion
these are not protocol errors, and are indicative. The following table enumerates server errors that can occur, where they occur, and the reason for the error.
```

a production deployment of OPAQUE: for instance, ensuring that there is a
server-authenticated channel over which OPAQUE registration and login is run.

# API Considerations

## Errors {#api-errors}

Some functions included in this specification are fallible. For example,
MACs exchanged during the AKE might fail verification, yielding an error. Additionally, protocol messages 
sent between client and server might not match their expected size, yielding a different error.
The following errors are documented for the OPRF-3DH composition in this specification and the recommended
configurations, and do not cover potential errors in other configurations. Participants MUST abort the
protocol when these errors occur.

### Deserialization errors {#deserialization-errors}

| Name                                 | Role   | Messages                  | Reason                                               |
|:-------------------------------------|:-------|:--------------------------|:-----------------------------------------------------|
| InvalidMessageLengthError            | Both   | All                       | The message length is invalid for the configuration. |

> | InvalidBlindedDataError              | Server | RegistrationRequest, Ke1  | Blinded data is an invalid point.                    |

Reviewer: ```suggestion
| InvalidBlindedDataError              | Server | RegistrationRequest, Ke1  | Blinded data is an invalid element.                    |
```

Reviewer: Is this raised in the spec? I guess it's implicit, but we should probably note that otherwise.

Author: What do you mean?

Author: Yes, it's implicit. How do you suggest we should specify that?

Author: Add a second table with the implicit exceptions that come from other specs?

> | InvalidEvaluatedDataError            | Client | RegistrationResponse, Ke2 | Invalid OPRF evaluation point.                       |

Reviewer: ```suggestion
| InvalidEvaluatedDataError            | Client | RegistrationResponse, Ke2 | Invalid OPRF key derivation.                       |
```
Also, it doesn't look like this is actually raised in the spec?

Author: Yes, many are implicit.

What about adding a dedicated table for implicit errors coming from other specs?

Reviewer: Can't we just add these alongside the existing errors? I think it'd also be good to note these errors alongside errors that OPAQUE generates on its own.

> | InvalidServerPublicKeyError          | Client | RegistrationResponse      | Invalid server public key.                           |

Reviewer: What does "invalid public key" mean here? (Same comment as below)

Author: To me, it's not a valid point in the group. Would another name be more suitable?

Author: Note that it's different from the deserialization: an input can correctly deserialize, but if the config dictates a different group, than the deserialized input would not be valid for that config

| InvalidClientPublicKeyError          | Server | Record                    | Invalid client public key.                           |
| InvalidClientEphemeralPublicKeyError | Server | Ke1                       | Invalid ephemeral client public key.                 |
| InvalidServerEphemeralPublicKeyError | Client | Ke2                       | Invalid ephemeral server public key.                 |

### Client errors {#client-errors}

The client can produce errors due to incorrect values in the messages it received. The following
table enumerates some of these errors, where they occur, and the reason for the error.

| Name                        | Stage                                                  | Reason                                                                   |
|:----------------------------|:-------------------------------------------------------|:-------------------------------------------------------------------------|

> | InvalidInputError           | CreateRegistrationRequest,<br/>CreateCredentialRequest | OPRF input deterministically maps to the group identity element.         |

Reviewer: ```suggestion
| InvalidInputError           | CreateRegistrationRequest, CreateCredentialRequest | OPRF input deterministically maps to the group identity element.         |
```

> | InvalidMaskedLengthError    | ClientFinish                                           | The length of the masked response is not = point length + envelope size. |

Reviewer: ```suggestion
| InvalidMaskedLengthError    | ClientFinish                                           | Invalid masked response length. |
```

| InvalidServerPublicKeyError | ClientFinish                                           | Invalid server public key in unmasked response.                          |

> | InvalidEnvelopeMACError     | ClientFinish                                           | Invalid envelope authentication tag.                                     |

Reviewer: ```suggestion
| InvalidEnvelopeMACError     | ClientFinish                                           | Envelope authentication check failed.                                     |
```

(And likewise below?)

### Server errors {#server-errors}

The following table enumerates server errors that can occur, where they occur, and the reason for the error.
In this case, servers can only generate a single protocol error.

| Name                     | Stage        | Reason                  |
|:-------------------------|:-------------|:------------------------|
| InvalidClientAkeMACError | ServerFinish | Invalid client AKE MAC. |

### Application errors

Implementations MUST verify certain input arguments and return an appropriate error. Note that
these are not protocol errors, and are indicative. The following table enumerates server errors that can occur, where they occur, and the reason for the error.

| Name                        | Role   | Stage      | Reason                                         |
|:----------------------------|:-------|:-----------|:-----------------------------------------------|
| InvalidServerSecretKeyError | Server | LoginInit  | Invalid server secret key.                     |
| InvalidServerPublicKeyError | Server | LoginInit  | Invalid server public key.                     |

> | InvalidOPRFSeedLengthError  | Server | LoginInit  | Invalid OPRF seed length.                      |

Reviewer: Why would this be thrown from the protocol, rather than enforced as a pre-condition to LoginInit? (Also, what is the LoginInit stage? It's not listed in the doc.)

Author: Yes, I wanted to discuss this in the PR-draft stage. As stated, implementations must observe these conditions, and can raise these errors, since these conditions are necessary for the protocol to execute correctly

Author: I suppose LoginInit comes from my implementation, my bad! Amending right now
